### PR TITLE
BIND backend: Several improvements for b2b-migrate

### PR DESCRIPTION
**Pull request by pieterlexis**
### Short description
While attempting to do a migration from the BIND backend to gsqlite3, several issues arose:

* No masters were set in the target db (#5807)
* Only the last master in the list of masters would be added to the target database
* The BIND backend was not fully aware of native zones

The last issue stemmed from an incomplete implementation in #5115.

This PR aims to fix all these issues.

Note: the b2b-migrate patch for setting the masters correctly and the patch to add the masters to the DomainInfo might need to be backported to 4.0.x.

### Checklist

I have:
- [x] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [x] compiled and tested this code
- [ ] included documentation (including possible behaviour changes)
- [ ] documented the code
- [ ] added or modified regression test(s)
- [ ] added or modified unit test(s)

### CODE REVIEW

File: modules/bindbackend/bindbackend2.cc [hunk 1]
```diff
@@ -353,6 +353,7 @@ void Bind2Backend::getAllDomains(vector<DomainInfo> *domains, bool include_disab
       di.zone=i->d_name;
       di.last_check=i->d_lastcheck;
       di.kind=i->d_masters.empty() ? DomainInfo::Master : DomainInfo::Slave; //TODO: what about Native?
```
Comment (cmouse) [inline modules/bindbackend/bindbackend2.cc]: `i->d_kind` instead?

File: modules/bindbackend/bindbackend2.cc [hunk 2]
```diff
@@ -302,7 +302,8 @@ void Bind2Backend::getUpdatedMasters(vector<DomainInfo> *changedDomains)
     ReadLock rl(&s_state_lock);
 
     for(state_t::const_iterator i = s_state.begin(); i != s_state.end() ; ++i) {
-      if(!i->d_masters.empty() && this->alsoNotify.empty() && i->d_also_notify.empty())
+      if(i->d_kind != DomainInfo::Slave ||
+          (!i->d_masters.empty() && this->alsoNotify.empty() && i->d_also_notify.empty()))
```
Comment (cmouse) [inline modules/bindbackend/bindbackend2.cc]: should these be removed or are Master domains with masters considered slaves?

**Comment by pieterlexis** [comment]
paging @mind04 for a review

File: modules/bindbackend/bindbackend2.cc (hunk 2 above)
Comment (pieterlexis) [reply modules/bindbackend/bindbackend2.cc]: Good question, I'd have to check what the expected behaviour is.

Comment (pieterlexis) [reply modules/bindbackend/bindbackend2.cc]: In the gsqlite3 backend, we `SELECT ... where type='MASTER'`. This also means that alsoNotify and d_masters don't even matter. We should not use NATIVE or SLAVE zones here to make the BIND backend behave consistent with the gsql backends. I'll make that so.

**Comment by mind04** [comment]
Two questions:

1. Was the interaction between master and slave checked? (notifies and freshness (soa) checks still working)
2. Is the behaviour of bind backed still compatible with real bind after this change for master/slave zones and also-notify?

**Comment by pieterlexis** [comment]
> Was the interaction between master and slave checked? (notifies and freshness (soa) checks still working)

Yes, this was tested and it works

> Is the behaviour of bind backed still compatible with real bind after this change for master/slave zones and also-notify?

The behaviour indeed differs from before (now that I re-read this code), fixing.

**Pull request merged**